File: legacy/src/main/rs/emulate/legacy/version/VersionListDecoder.java
package rs.emulate.legacy.version;

import com.google.common.collect.ImmutableList;
import rs.emulate.legacy.archive.Archive;
import rs.emulate.legacy.archive.ArchiveEntry;
import rs.emulate.shared.util.DataBuffer;

import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Decodes ArchiveEntry objects from the version {@link Archive} containing file version data.
 *
 * @author Major
 */
public final class VersionListDecoder {

	/**
	 * The version Archive.
	 */
	private final Archive versions;

	/**
	 * Creates the VersionListDecoder.
	 *
	 * @param versions The {@link Archive} containing the version data.
	 */
	public VersionListDecoder(Archive versions) {
		this.versions = versions;
	}

	/**
	 * Decodes the file {@link VersionList}s.
	 *
	 * @return The {@link List} of file VersionLists.
	 * @throws FileNotFoundException If any of the {@link ArchiveEntry} names could not be found.
	 */
	public List<VersionList> decode() throws FileNotFoundException {
		List<VersionList> lists = new ArrayList<>(VersionList.VERSION_ENTRY_NAMES.length);

		for (int type = 0; type < VersionList.VERSION_ENTRY_NAMES.length; type++) {
			String name = VersionList.VERSION_ENTRY_NAMES[type];
			ArchiveEntry entry = versions.getEntry(name);
			DataBuffer data = entry.getBuffer();

			int count = data.limit() / Short.BYTES;
			int[] versions = new int[count];
			Arrays.setAll(versions, index -> data.getUnsignedShort());

			lists.add(new VersionList(versions));
		}

		return ImmutableList.copyOf(lists);
	}

}

File: legacy/src/main/kotlin/rs/emulate/legacy/version/VersionListDecoder.kt
package rs.emulate.legacy.version

import rs.emulate.legacy.archive.Archive
import rs.emulate.legacy.archive.ArchiveEntry
import java.io.FileNotFoundException
import java.util.ArrayList
import java.util.Arrays

/**
 * Decodes ArchiveEntry objects from the version [Archive] containing file version data.
 *
 * @param versions The [Archive] containing the version data.
 */
class VersionListDecoder(private val versions: Archive) {

    /**
     * Decodes the file [VersionList]s.
     * @throws FileNotFoundException If any of the [ArchiveEntry] names could not be found.
     */
    fun decode(): List<VersionList> {
        val lists = ArrayList<VersionList>(VersionList.VERSION_ENTRY_NAMES.size)

        for (type in VersionList.VERSION_ENTRY_NAMES.indices) {
            val name = VersionList.VERSION_ENTRY_NAMES[type]
            val entry = versions[name]
            val data = entry.buffer

            val count = data.readableBytes() / java.lang.Short.BYTES
            val versions = IntArray(count)
            Arrays.setAll(versions) { data.readUnsignedShort() }

            lists += VersionList(versions)
        }

        return lists
    }

}
